New error with GCC7

Building the project with GCC7, I see:

[   59s] cl_pidfile.c:172:29: error: '__builtin___snprintf_chk' output may be truncated before the last format character [-Werror=format-truncation=]
[   59s]   snprintf(buf, sizeof(buf), "%*ld\n", LOCKSTRLEN-1, mypid);
[   59s]                              ^~~~~~~~
[   59s] In file included from /usr/include/stdio.h:939:0,
[   59s]                  from cl_pidfile.c:18:
[   59s] /usr/include/bits/stdio2.h:64:10: note: '__builtin___snprintf_chk' output between 12 and 13 bytes into a destination of size 12
[   59s]    return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
[   59s]           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   59s]         __bos (__s), __fmt, __va_arg_pack ());
[   59s]         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Which can be easily fixed by increasing LOCKSTRLEN from 11 to 12, which is enough even for 64-bit long numbers. I can prepare pull request whether it's preferred way how to fix the issue?

On Mon, Mar 20, 2017 at 06:17:59AM -0700, marxin wrote:
> Building the project with GCC7, I see:
> 
> [   59s] cl_pidfile.c:172:29: error: '__builtin___snprintf_chk' output may be truncated before the last format character [-Werror=format-truncation=]
> [   59s]   snprintf(buf, sizeof(buf), "%*ld\n", LOCKSTRLEN-1, mypid);
> [   59s]                              ^~~~~~~~
> [   59s] In file included from /usr/include/stdio.h:939:0,
> [   59s]                  from cl_pidfile.c:18:
> [   59s] /usr/include/bits/stdio2.h:64:10: note: '__builtin___snprintf_chk' output between 12 and 13 bytes into a destination of size 12
> [   59s]    return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
> [   59s]           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> [   59s]         __bos (__s), __fmt, __va_arg_pack ());
> [   59s]         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> Which can be easily fixed by increasing LOCKSTRLEN from 11 to
> 12, which is enough even for 64-bit long numbers. I can prepare
> pull request whether it's preferred way how to fix the issue?

Yes, please. LOCKSTRLEN is used only for automatic variables,
hence we don't need to think about the library compatibility
issues.

There's just one awkward test in ./lib/clplumbing/cl_pidfile.c:

```
137         if (fstat(fd, &sbuf) >= 0 && sbuf.st_size < LOCKSTRLEN) {
138             sleep(1); /* if someone was about to create one,
139                    * give'm a sec to do so
```

If someone's unlucky there should only be an extra 1 second
sleep.

--- a/lib/clplumbing/cl_pidfile.c
+++ b/lib/clplumbing/cl_pidfile.c
@@ -71,7 +71,7 @@
 
 
 /* This is what the FHS standard specifies for the size of our lock file */
-#define	LOCKSTRLEN	11
+#define	LOCKSTRLEN	12
 #include <clplumbing/cl_log.h>
 int IsRunning(long pid)
 {

--- a/lib/stonith/st_ttylock.c
+++ b/lib/stonith/st_ttylock.c
@@ -125,7 +125,7 @@ st_ttyunlock(const char *serial_device)
 }
 
 /* This is what the FHS standard specifies for the size of our lock file */
-#define	LOCKSTRLEN	11
+#define	LOCKSTRLEN	12
 
 static int
 DoLock(const char * prefix, const char *lockname)

